code/Control/UserDefinedFormController.php

```diff
@@ -350,6 +358,10 @@ public function process($data, $form)
                 // This string substitution works for both HTML and plain text emails.
                 // $recipient->getEmailBodyContent() will retrieve the relevant version of the email
                 $emailData['Body'] = SSViewer::execute_string($recipient->getEmailBodyContent(), $mergeFields);
+                // only include visible fields if recipient visibility flag is set
+                if ((bool) $recipient->HideInvisibleFields) {
+                    $emailData['Fields'] = $visibleSubmittedFields;
```

Comment:
```suggestion
                    $emailData['Fields'] = $emailData['Fields']->filter('Displayed', true);
```

Reply:
This only works for the first recipient. After that `$emailData['Fields']` is changed and can't be set back to all fields. That's why we need two array to store both versions of fields.

Reply:
Ok, fair enough. Basically I want to avoid pushing each item separately into a new array, but we can use a variable and assign the filtered list to it for reuse.

Reply:
Fair enough. That's changed now.